File: submissions/climbing-stairs.c
#include <string.h>


int climbStairs(int n) {
    
    // Create an array to store results
    int results[46];
    memset(results, 0, sizeof(results));

    // Run the function
    int output = climbStairsRecursive(n, results);

    return output;
}


int climbStairsRecursive(int n, int results[]){

    // Base case: No more steps
    if (n == 0) {
        return 1;
    }
    // Base case: Value has been computed previously
    if (results[n] != 0) {
        return results[n];
    }

    int output = 0;

    // Call function recursively, trying one and two steps
    if (n >= 1) {
        output += climbStairsRecursive(n - 1, results);
    }
    if (n >= 2) {
        output += climbStairsRecursive(n - 2, results);
    }

    // Store results to avoid recomputation
    results[n] = output;

    return output;
}

File: submissions/climbing-stairs.py
def _climb_stairs(n: int, memo: dict = None):
    
    if n not in memo:
    
        output = 0
        if n >= 2:
            output += _climb_stairs(n - 2, memo)
        if n >= 1:
            output += _climb_stairs(n - 1, memo)

        memo[n] = output

    return memo[n]


class Solution:
    def climbStairs(self, n: int) -> int:
        # Initialize the memo with the base case
        memo = {0: 1}
        return _climb_stairs(n, memo)
